concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,194 @@
+//! cleanup and advanced allocation schemes is removed.
+//!
+//! The bump allocator here has two extra features:
+//!  - It keeps track of the number of active allocations in the `allocations`
+//!    field and resets the `next` pointer to `MIN_PTR_ADDR` if there are no
+//!    active allocations, thus allowing the reuse of memory.
+//!  - On deallocations it checks whether it is the very last memory block
+//!    handed out that is deallocated. If it is, then it moves the `next`
+//!    pointer back to the beginning of that block. Thus also allowing memory to
+//!    be reused.
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    cell::UnsafeCell,
+    ptr,
+};
+
+/// The byte size of Wasm pages.
+const PAGE_SIZE: usize = 65536;
+/// The minimum pointer address used. Address `0` cannot be used, as that
+/// corresponds to a null pointer, which indicates errors.
+const MIN_PTR_ADDR: usize = 1;
+
+/// A number of WebAssembly memory pages.
+#[derive(Eq, PartialEq, Copy, Clone)]
+struct PageCount(usize);
+
+impl PageCount {
+    /// The size in bytes. I.e. `PageCount * PAGE_SIZE`.
+    fn size_in_bytes(&self) -> usize { self.0 * PAGE_SIZE }
+}
+
+/// The number of pages returned from `memory_grow` to indicate out of memory.
+const ERROR_PAGE_COUNT: PageCount = PageCount(usize::MAX);
+
+/// This is an invalid implementation of [`Sync`].
+/// The [`BumpAllocator`] cannot be safely used from multiple threads, but that
+/// is OK since it won't be in Wasm. This [`Sync`] implementation is required
+/// for defining it as the global allocator.
+unsafe impl Sync for BumpAllocator {}
+
+pub struct BumpAllocator {
+    /// The pointer to the next memory to hand out.
+    next:        UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to
+    /// `MIN_PTR_ADDR` if there are no more active allocations.
+    allocations: UnsafeCell<usize>,
+    /// Stores the last address given out, and allows for a small improvement
+    /// over the always-increasing bump allocator. Namely that if an item
+    /// `a` is allocated and deallocated without any allocations in between,
+    /// then the memory of `a` is reused for the next allocation.
+    last_alloc:  UnsafeCell<usize>,
+}
+
+impl BumpAllocator {
+    /// Create a new [`BumpAllocator`].
+    ///
+    /// This must be a `const` method for it to be used as a global allocator,
+    /// where it is called staticly. Since we cannot know the actual size of
+    /// the heap, `heap_end` is initialized to `0` and updated on the first
+    /// allocation.
+    ///
+    /// # Safety
+    /// - Can only be used in single-threaded environments. The [`Sync`]
+    ///   implementation is only invalid and is only there to support setting
+    ///   this as the global allocator.
```

Comment:
Are we sure we are not breaking some invariants with implementing this?

I'm guessing not

Reply:
No, let's do this.